### PR TITLE
[TGUI] Upgrades Bots

### PR DESCRIPTION
**Pull request by Kyep**
## What Does This PR Do & Why

TGUI Conversions (because the old interface was atrocious):
- Converted Securitron (and its variants: beepsky, pingsky, griefsky, ofitser, buzzky, armsky) to TGUI
- Converted ED209 (and its variants: redtag, bluetag, syndisentrybot) to TGUI
- Converted Cleanerbots to TGUI

Other SecBot Changes:
- Secbots (securitron, ED209) that have "check weapons" turned on now arrest people for having unauthorized weapons in their belt or suit storage slots. Previously, having them in your hand would get you arrested, but having them on your belt or suit would not, which was very inconsistent seeing as all were clearly visible on you. Now they're handled the same way. Common toy weapons (e.g: lasertag guns), and default job weapons (e.g: nullrods, KAs), are always considered authorized.
- Fixed beepsky/bots failing to recognize some obviously deadly weapons. Examples include dual energy swords (even though single energy swords were recognized), katanas, spears, lava/storm staff weapons, and bone axes.

Turret Changes:
- Turrets are no longer racist, specifically they no longer open fire on a humanoid but non-human (like an IPC, vox or tajaran) when they have a weapon on their belt. They used to do this, even though they wouldn't fire on a human in the exact same situation.
- No longer use istype() checks for their weapon checks, instead using the same item.needs_permit check that bots use, which in addition to being way more consistent, and reducing istype() checks, also means they won't for example shoot a miner for having a KA, or someone for having a toy gun. Which they used to do.
- Now permit people with weapons permits to carry weapons, as opposed to the old system where they'd ignore weapon permits entirely and instead think you had a weapons permit if and only if you had access to the turret controls. That made weapon permits totally useless as far as turrets went. All that is now fixed.

## Images of changes
Secbot (ie: beepsky) / ED209 interface:
new_beepsky
Note that the "hack" option only shows up for AIs/aghosts.
Cleanerbot interface:
new_cleanbot

## Changelog
:cl: Kyep
tweak: Converted secbots (including beepsky), ED209s, and cleanerbots to TGUI. Their interface looks much nicer now.
tweak: Made secbot/ED209 weapon handling more consistent, most notably they no longer semi-ignore weapons on belts/suits, all spears, and putting two eswords together no longer causes the bot to fail to recognize the resulting double-sword as a weapon.
tweak: Turrets are no longer racist against every race except basic human, shooting them for things they would not shoot a human for.
tweak: Turrets no longer shoot people for having toy guns, or for having a proper weapon despite having a weapons permit.
/:cl:

### CODE REVIEW

**Comment by AffectedArc07** [comment]
Classified as tweak not refactor since its a minor functionality change with the belt slot thing

File: code/modules/mob/living/simple_animal/bot/bot.dm [hunk 1]
```diff
@@ -871,29 +871,37 @@ Pass a positive integer as an argument to override a bot's default speed.
...
+/mob/living/simple_animal/bot/proc/canhack(mob/M)
+	return ((issilicon(M) && !emagged) || M.can_admin_interact())
+
+/mob/living/simple_animal/bot/proc/handle_hacking(mob/M) // refactored out of Topic/ to allow re-use by TGUIs
+	if(!canhack(M))
+		return
+	if(emagged != 2)
+		emagged = 2
+		hacked = 1
+		locked = 1
```
Comment (AffectedArc07) [inline code/modules/mob/living/simple_animal/bot/bot.dm:890]:
```suggestion
		hacked = TRUE
		locked = TRUE
```

File: code/modules/mob/living/simple_animal/bot/bot.dm [hunk 2]
```diff
@@ -871,29 +871,37 @@ Pass a positive integer as an argument to override a bot's default speed.
...
+		to_chat(M, "<span class='warning'>[text_hack]</span>")
+		show_laws()
+		bot_reset()
+		add_attack_logs(M, src, "Hacked")
+	else if(!hacked)
+		to_chat(M, "<span class='userdanger'>[text_dehack_fail]</span>")
+	else
+		emagged = 0
+		hacked = 0
```
Comment (AffectedArc07) [inline code/modules/mob/living/simple_animal/bot/bot.dm:899]:
```suggestion
		emagged = FALSE
		hacked = FALSE
```

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 3]
```diff
@@ -190,29 +190,60 @@
...
-Status: []<BR>
-Behaviour controls are [locked ? "locked" : "unlocked"]<BR>
-Maintenance panel panel is [open ? "opened" : "closed"]"},
-text("<A href='?src=[UID()];power=1'>[on ? "On" : "Off"]</A>"))
-	if(!locked || issilicon(user) || user.can_admin_interact())
-		dat += text({"<BR>Cleans Blood: []<BR>"}, text("<A href='?src=[UID()];operation=blood'>[blood ? "Yes" : "No"]</A>"))
-		dat += text({"<BR>Patrol station: []<BR>"}, text("<A href='?src=[UID()];operation=patrol'>[auto_patrol ? "Yes" : "No"]</A>"))
-	return dat
-
-/mob/living/simple_animal/bot/cleanbot/Topic(href, href_list)
```
Comment (AffectedArc07) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:208]: Please test this with PDA bot controls. I believe those use radio signal `topic`s in the same way that the AI door control did, so be sure.

File: code/modules/mob/living/simple_animal/bot/secbot.dm [hunk 4]
```diff
@@ -111,61 +111,73 @@
...
-"<A href='?src=[UID()];operation=idcheck'>[idcheck ? "Yes" : "No"]</A>",
-"<A href='?src=[UID()];operation=weaponscheck'>[weaponscheck ? "Yes" : "No"]</A>",
-"<A href='?src=[UID()];operation=ignorerec'>[check_records ? "Yes" : "No"]</A>",
-"<A href='?src=[UID()];operation=switchmode'>[arrest_type ? "Detain" : "Arrest"]</A>",
-"<A href='?src=[UID()];operation=declarearrests'>[declare_arrests ? "Yes" : "No"]</A>",
-"<A href='?src=[UID()];operation=patrol'>[auto_patrol ? "On" : "Off"]</A>" )
-
-	return	dat
-
-/mob/living/simple_animal/bot/secbot/Topic(href, href_list)
```
Comment (AffectedArc07) [inline code/modules/mob/living/simple_animal/bot/secbot.dm:149]: Same here

File: tgui/packages/tgui/interfaces/BotClean.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    noaccess,
+    maintpanel,
+    on,
+    autopatrol,
+    canhack,
+    emagged,
+    remote_disabled,
+    painame,
+
+    cleanblood,
```
Comment (AffectedArc07) [inline tgui/packages/tgui/interfaces/BotClean.js:19]:
```suggestion
    painame,
    cleanblood,
```

File: tgui/packages/tgui/interfaces/BotClean.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            <LabeledList.Item label="Safety System">
+              <Box color={emagged ? 'bad' : 'good'}>
+                {emagged ? "DISABLED!" : "Enabled"}
+              </Box>
+            </LabeledList.Item>
+            {!!canhack && (
+              <LabeledList.Item label="Hacking">
+                <Button
+                  icon="terminal"
+                  content={emagged ? "UnHack" : "Hack"}
```
Comment (AffectedArc07) [inline tgui/packages/tgui/interfaces/BotClean.js:61]:
```suggestion
                  content={emagged ? "Unhack" : "Hack"}
```
Maybe even `Restore` or `Revert`, something about "Unhack" just sounds off, and the spelling of UnHack is definitely off.

File: tgui/packages/tgui/interfaces/BotClean.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            fluid
+            checked={cleanblood}
+            content="Clean Blood"
+            disabled={noaccess}
+            onClick={() => act('blood')} />
+        </Section>
+        {painame && (
+          <Section title="pAI">
+            <Button
+              icon="eject"
```
Comment (AffectedArc07) [inline tgui/packages/tgui/interfaces/BotClean.js:88]: Make the pAI button `fluid` as well if all the others are

File: tgui/packages/tgui/interfaces/BotSecurity.js [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    canhack,
+    emagged,
+    remote_disabled,
+    painame,
+
+    check_id,
+    check_weapons,
+    check_warrant,
+
+    arrest_mode,
```
Comment (AffectedArc07) [inline tgui/packages/tgui/interfaces/BotSecurity.js:23]:
```suggestion
    painame,
    check_id,
    check_weapons,
    check_warrant,
    arrest_mode,
```

File: tgui/packages/tgui/interfaces/BotSecurity.js [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+            fluid
+            checked={arrest_declare}
+            content="Announce Arrests On Radio"
+            disabled={noaccess}
+            onClick={() => act('arrdeclare')} />
+        </Section>
+        {painame && (
+          <Section title="pAI">
+            <Button
+              icon="eject"
```
Comment (AffectedArc07) [inline tgui/packages/tgui/interfaces/BotSecurity.js:119]: Same here

**Review by AffectedArc07 — CHANGES_REQUESTED** [review]
First pass

File: code/modules/mob/living/simple_animal/bot/secbot.dm (hunk 4 above)
Comment (Kyep) [reply code/modules/mob/living/simple_animal/bot/secbot.dm:149]: Confirmed that both secbots and ED209s still work with the PDA app.

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm (hunk 3 above)
Comment (Kyep) [reply code/modules/mob/living/simple_animal/bot/cleanbot.dm:208]: There aren't any PDA controls for cleanbots. The nearest thing is the Custodial Locator PDA app the Janitor gets, but even there, all it does is show the bot's co-ords... it doesn't allow remote control of the bots.

**Review by Necaladun — APPROVED** [review]
yeah fine but it would be nice to have a toggle to add the racism back in ;)

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 10]
```diff
@@ -190,29 +190,60 @@
...
+/mob/living/simple_animal/bot/cleanbot/tgui_data(mob/user)
+	var/list/data = list(
+		"locked" = locked, // controls, locked or not
+		"noaccess" = topic_denied(user), // does the current user have access? admins, silicons etc can still access bots with locked controls
+		"maintpanel" = open,
+		"on" = on,
+		"autopatrol" = auto_patrol,
+		"painame" = paicard ? paicard.pai.name : null,
+
+		"canhack" = canhack(user),
```
Comment (AffectedArc07) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:211]:
```suggestion
		"painame" = paicard ? paicard.pai.name : null,
		"canhack" = canhack(user),
```

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 11]
```diff
@@ -190,29 +190,60 @@
...
+		"emagged" = emagged, // this is an int, NOT a boolean
+		"remote_disabled" = remote_disabled,
+
+		"cleanblood" = blood
```
Comment (AffectedArc07) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:215]:
```suggestion
		"remote_disabled" = remote_disabled,
		"cleanblood" = blood
```

File: code/modules/mob/living/simple_animal/bot/ed209bot.dm [hunk 12]
```diff
@@ -102,64 +102,80 @@
...
+/mob/living/simple_animal/bot/ed209/tgui_data(mob/user)
+	var/list/data = list(
+		"locked" = locked, // controls, locked or not
+		"noaccess" = topic_denied(user), // does the current user have access? admins, silicons etc can still access bots with locked controls
+		"maintpanel" = open,
+		"on" = on,
+		"autopatrol" = auto_patrol,
+		"painame" = paicard ? paicard.pai.name : null,
+
+		"canhack" = canhack(user),
```
Comment (AffectedArc07) [inline code/modules/mob/living/simple_animal/bot/ed209bot.dm:123]:
```suggestion
		"painame" = paicard ? paicard.pai.name : null,
		"canhack" = canhack(user),
```

File: code/modules/mob/living/simple_animal/bot/ed209bot.dm [hunk 13]
```diff
@@ -102,64 +102,80 @@
...
+		"emagged" = emagged, // this is an int, NOT a boolean
+		"remote_disabled" = remote_disabled,
+
+		"check_id" = idcheck,
```
Comment (AffectedArc07) [inline code/modules/mob/living/simple_animal/bot/ed209bot.dm:127]:
```suggestion
		"remote_disabled" = remote_disabled,
		"check_id" = idcheck,
```

File: code/modules/mob/living/simple_animal/bot/ed209bot.dm [hunk 14]
```diff
@@ -102,64 +102,80 @@
...
+		"check_weapons" = weaponscheck,
+		"check_warrant" = check_records,
+
+		"arrest_mode" = arrest_type, // detain or arrest
```
Comment (AffectedArc07) [inline code/modules/mob/living/simple_animal/bot/ed209bot.dm:131]:
```suggestion
		"check_warrant" = check_records,
		"arrest_mode" = arrest_type, // detain or arrest
```

File: code/modules/mob/living/simple_animal/bot/secbot.dm [hunk 15]
```diff
@@ -111,61 +111,74 @@
...
+/mob/living/simple_animal/bot/secbot/tgui_data(mob/user)
+	var/list/data = list(
+		"locked" = locked, // controls, locked or not
+		"noaccess" = topic_denied(user), // does the current user have access? admins, silicons etc can still access bots with locked controls
+		"maintpanel" = open,
+		"on" = on,
+		"autopatrol" = auto_patrol,
+		"painame" = paicard ? paicard.pai.name : null,
+
+		"canhack" = canhack(user),
```
Comment (AffectedArc07) [inline code/modules/mob/living/simple_animal/bot/secbot.dm:136]:
```suggestion
		"painame" = paicard ? paicard.pai.name : null,
		"canhack" = canhack(user),
```

File: code/modules/mob/living/simple_animal/bot/secbot.dm [hunk 16]
```diff
@@ -111,61 +111,74 @@
...
+		"emagged" = emagged, // this is an int, NOT a boolean
+		"remote_disabled" = remote_disabled,
+
+		"check_id" = idcheck,
```
Comment (AffectedArc07) [inline code/modules/mob/living/simple_animal/bot/secbot.dm:140]:
```suggestion
		"remote_disabled" = remote_disabled,
		"check_id" = idcheck,
```

File: code/modules/mob/living/simple_animal/bot/secbot.dm [hunk 17]
```diff
@@ -111,61 +111,74 @@
...
+		"check_weapons" = weaponscheck,
+		"check_warrant" = check_records,
+
+		"arrest_mode" = arrest_type, // detain or arrest
```
Comment (AffectedArc07) [inline code/modules/mob/living/simple_animal/bot/secbot.dm:144]:
```suggestion
		"check_warrant" = check_records,
		"arrest_mode" = arrest_type, // detain or arrest
```

**Review by AffectedArc07 — CHANGES_REQUESTED** [review]
Few styling nitpicks

**Review by AffectedArc07 — APPROVED** [review]

**Comment by Jungling** [comment]
TGUI hype

**Pull request merged**